stock_routing_operation/models/stock_move.py

```diff
@@ -0,0 +1,403 @@
+                new_move_per_location[routing_location.id].append(new_move_id)
+
+        new_moves = self.browse(chain.from_iterable(new_move_per_location.values()))
+        return self | new_moves
+
+    def _apply_routing_rule_pull(self):
+        """Apply routing operations
+
+        When a move has a routing operation configured on its location and the
+        destination of the move does not match the destination of the routing
+        operation, this method updates the move's destination and it's picking
+        type with the routing operation ones and creates a new chained move
+        after it.
+        """
+        pickings_to_check_for_emptiness = self.env["stock.picking"]
+        for move in self:
+            routing_rule = move.routing_rule_id
+            if not routing_rule.method == "pull":
+                continue
+
+            if move.picking_id.picking_type_id == routing_rule.picking_type_id:
+                # already correct
+                continue
+
+            # we expect all the lines to go to the same destination for
+            # pull routing rules
+            original_destination = move.location_dest_id
+            current_picking_type = move.picking_id.picking_type_id
+            move.with_context(
+                __applying_routing_rule=True
+            ).location_id = routing_rule.location_src_id
+            move.picking_type_id = routing_rule.picking_type_id
+            if self.env["stock.location"].search(
+                [
+                    ("id", "=", routing_rule.location_dest_id.id),
+                    ("id", "child_of", move.location_dest_id.id),
+                ]
+            ):
+                # The destination of the move, as a parent of the destination
+                # of the routing, goes to the correct place, but is not precise
+                # enough: set the new destination to match the rule's one.
+                # The source of the dest. move will be changed to match it,
+                # which may reapply a new routing rule on the dest. move.
+                move._routing_pull_switch_destination(routing_rule)
+
+            elif not self.env["stock.location"].search(
+                [
+                    ("id", "=", routing_rule.location_dest_id.id),
+                    ("id", "parent_of", move.location_dest_id.id),
+                ]
+            ):
+                # The destination of the move is unrelated (nor identical, nor
+                # a parent or a child) to the routing destination: in this case
+                # we have to add a routing move before the current move to
+                # route the goods in the correct place
+                move._routing_pull_insert_move(
+                    routing_rule, current_picking_type, original_destination
+                )
+
+            pickings_to_check_for_emptiness |= move.picking_id
+            move._assign_picking()
+            move._action_assign()
```

Comment:
We should not assign the move here because it will be just after when the _action_assign's super will be called?

Reply:
It has to be reserved at this point. If the routing has to split a move because it was partially available, you want to control which move will be reserved: the one on which you applied the routing which is this one. So we assign it first. Note that the second call to `_action_assign()` will do nothing because we are guaranteed to have a state "assigned" on this move and the method exclude this state.

Reply:
Added a comment